"""
æµ‹è¯• gRPC å®¢æˆ·ç«¯
éªŒè¯ç«¯åˆ°ç«¯æµå¼é€šä¿¡
"""
import asyncio
import grpc
from loguru import logger

from app.gen.agent.v1 import agent_service_pb2, agent_service_pb2_grpc


async def test_stream_chat():
    """
    æµ‹è¯•æµå¼èŠå¤©åŠŸèƒ½
    """
    logger.info("Starting gRPC StreamChat test...")

    # è¿æ¥åˆ° gRPC æœåŠ¡å™¨
    async with grpc.aio.insecure_channel('localhost:50051') as channel:
        stub = agent_service_pb2_grpc.AgentServiceStub(channel)

        # æ„å»ºè¯·æ±‚
        request = agent_service_pb2.ChatRequest(
            user_id="test_user_123",
            session_id="test_session_456",
            message="ä½ å¥½ï¼Œè¯·ä»‹ç»ä¸€ä¸‹ä½ è‡ªå·±",
            user_profile=agent_service_pb2.UserProfile(
                nickname="æµ‹è¯•åŒå­¦",
                timezone="Asia/Shanghai",
                language="zh-CN",
                is_pro=False
            ),
            config=agent_service_pb2.ChatConfig(
                model="",
                temperature=0.7,
                max_tokens=500,
                tools_enabled=False
            ),
            request_id="test_req_001"
        )

        # æ·»åŠ  metadata
        metadata = (
            ("user-id", "test_user_123"),
            ("x-trace-id", "test_trace_001"),
        )

        try:
            logger.info("Sending StreamChat request...")
            logger.info(f"Request: user_id={request.user_id}, message={request.message}")

            response_count = 0
            full_response = ""

            # æ¥æ”¶æµå¼å“åº”
            async for response in stub.StreamChat(request, metadata=metadata):
                response_count += 1

                # å¤„ç†ä¸åŒç±»å‹çš„å“åº”
                if response.HasField("delta"):
                    print(response.delta, end="", flush=True)
                    full_response += response.delta

                elif response.HasField("status_update"):
                    status = response.status_update
                    state_name = agent_service_pb2.AgentStatus.State.Name(status.state)
                    logger.info(f"\nğŸ“ Status: {state_name} - {status.details}")

                elif response.HasField("full_text"):
                    logger.info(f"\nâœ… Full response received: {len(response.full_text)} chars")

                elif response.HasField("error"):
                    error = response.error
                    logger.error(f"\nâŒ Error: [{error.code}] {error.message}")

                elif response.HasField("usage"):
                    usage = response.usage
                    logger.info(f"\nğŸ“Š Usage: {usage.total_tokens} tokens")

            print("\n")  # æ¢è¡Œ
            logger.success(f"âœ… StreamChat completed! Received {response_count} chunks")
            logger.info(f"ğŸ“ Full response length: {len(full_response)} chars")

            return True

        except grpc.RpcError as e:
            logger.error(f"âŒ gRPC error: {e.code()} - {e.details()}")
            return False
        except Exception as e:
            logger.error(f"âŒ Unexpected error: {e}", exc_info=True)
            return False


async def test_retrieve_memory():
    """
    æµ‹è¯•è®°å¿†æ£€ç´¢åŠŸèƒ½
    """
    logger.info("\nStarting gRPC RetrieveMemory test...")

    async with grpc.aio.insecure_channel('localhost:50051') as channel:
        stub = agent_service_pb2_grpc.AgentServiceStub(channel)

        request = agent_service_pb2.MemoryQuery(
            user_id="test_user_123",
            query_text="é«˜ç­‰æ•°å­¦ æé™",
            limit=5,
            min_score=0.7,
            hybrid_alpha=0.8
        )

        try:
            logger.info("Sending RetrieveMemory request...")
            response = await stub.RetrieveMemory(request)

            logger.info(f"âœ… Found {response.total_found} memory items")
            for idx, item in enumerate(response.items, 1):
                logger.info(f"  {idx}. Score: {item.score:.3f} - {item.content[:50]}...")

            return True

        except grpc.RpcError as e:
            logger.error(f"âŒ gRPC error: {e.code()} - {e.details()}")
            return False


async def main():
    """
    è¿è¡Œæ‰€æœ‰æµ‹è¯•
    """
    logger.info("=" * 70)
    logger.info("ğŸ§ª Sparkle AI Agent gRPC Client Test Suite")
    logger.info("=" * 70)

    # æµ‹è¯• StreamChat
    test1_success = await test_stream_chat()

    # ç­‰å¾…ä¸€ä¸‹
    await asyncio.sleep(1)

    # æµ‹è¯• RetrieveMemory
    test2_success = await test_retrieve_memory()

    # æ€»ç»“
    logger.info("\n" + "=" * 70)
    logger.info("ğŸ“Š Test Results:")
    logger.info(f"  StreamChat:      {'âœ… PASS' if test1_success else 'âŒ FAIL'}")
    logger.info(f"  RetrieveMemory:  {'âœ… PASS' if test2_success else 'âŒ FAIL'}")
    logger.info("=" * 70)

    if test1_success and test2_success:
        logger.success("ğŸ‰ All tests passed!")
        return 0
    else:
        logger.error("âŒ Some tests failed")
        return 1


if __name__ == '__main__':
    exit_code = asyncio.run(main())
    exit(exit_code)
